solidity/contracts/BancorNetwork.sol

```diff
@@ -450,6 +405,191 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
+      * @param _smartToken  smart token of the first conversion step
+      * @param _amount      amount to convert from, in the source token
+    */
+    function handleSourceToken(IERC20Token _sourceToken, ISmartToken _smartToken, uint256 _amount) private {
+        bool isSourceETH = etherTokens[_sourceToken];
+
+        // validate msg.value
+        if (msg.value > 0)
+            require(msg.value == _amount && isSourceETH);
+        else
+            require(msg.value == 0);
+
+        IBancorConverter firstConverter = IBancorConverter(_smartToken.owner());
+        bool isNewerConverter = isV27OrHigherConverter(firstConverter);
+
+        // ETH / EtherToken
+        if (isSourceETH) {
+            // handle ETH
+            if (msg.value > 0) {
+                // EtherToken converter - deposit the ETH into the EtherToken
+                // note that it can still be a non ETH converter if the path is wrong
+                // but such conversion will simply revert
+                if (!isNewerConverter) {
+                    IEtherToken etherToken = getConverterEtherTokenAddress(firstConverter);
+                    IEtherToken(etherToken).deposit.value(msg.value)();
+                }
+            }
+            // handle EtherToken
+            else {
+                // claim the tokens
+                safeTransferFrom(_sourceToken, msg.sender, this, _amount);
+
+                // ETH converter - withdraw the ETH
+                if (isNewerConverter)
+                    IEtherToken(_sourceToken).withdraw(_amount);
+            }
+        }
+        // other ERC20 token
+        else {
+            // newer converter - transfer the tokens from the sender directly to the converter
+            // otherwise claim the tokens
+            if (isNewerConverter)
+                safeTransferFrom(_sourceToken, msg.sender, firstConverter, _amount);
+            else
+                safeTransferFrom(_sourceToken, msg.sender, this, _amount);
+        }
+    }
+
+    /**
+      * @dev handles the conversion target token if the network still holds it at the end of the conversion
+      * 
+      * @param _data        conversion data, see ConversionStep struct above
+      * @param _amount      conversion return amount, in the target token
+      * @param _beneficiary wallet to receive the conversion result
+    */
+    function handleTargetToken(ConversionStep[] _data, uint256 _amount, address _beneficiary) private {
+        ConversionStep memory stepData = _data[_data.length - 1];
+
+        // network contract doesn't hold the tokens, do nothing
+        if (stepData.beneficiary != address(this))
+            return;
+
+        IERC20Token targetToken = stepData.targetToken;
+
+        // ETH / EtherToken
+        if (etherTokens[targetToken]) {
+            // newer converter should send ETH directly to the beneficiary
+            assert(!stepData.isV27OrHigherConverter);

```

Comment:
I want to make sure that the assertion here is right:
If the user inputs a path which starts with EtherToken, then the owner of this EtherToken will most definitely be an older converter. Is that corrcet?
Or can the owner be a new converter?
In which case, even if you still don't want to support this scenario, you need to use `require`, because it's a legitimate user-input (not an "unreachable" scenario).

Reply:
This is the target token handling - not the source, so it's unrelated to the start of the path.
We only get to this case if the last conversion in the path sent the tokens to the network.
If the last conversion ends with ETH -
- New converter will always send that ETH to the target user (so these ETH tokens can never be currently held by the network in this case)
- Old converter will always send that ETH to the network